---
title: 'Practical: basic stats for peak calling'
author: "Jacques van Helden"
date: '`r Sys.Date()`'
output:
  slidy_presentation:
    fig_caption: yes
    fig_height: 6
    fig_width: 7
    footer: UNAM LCG BEII - Jacques van Helden - Introduction and contents
    highlight: tango
    incremental: no
    keep_md: no
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
  ioslides_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 6
    fig_width: 7
    fonttheme: structurebold
    footer: UNAM LCG BEII - Jacques van Helden - Introduction and contents
    highlight: tango
    incremental: no
    keep_md: no
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
  beamer_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 6
    fig_width: 7
    fonttheme: structurebold
    highlight: tango
    incremental: no
    keep_tex: no
    slide_level: 2
    theme: Montpellier
    toc: yes
font-import: http://fonts.googleapis.com/css?family=Risque
font-family: Garamond
address: Lab.Technological Advances for Genomics and Clinics, Inserm Unit U1090
transition: linear
---

## Peak-calling: Carl's question

<center>![peak-calling question](images/peak-calling_the_question.png)</center>


## Defining the data directory

Open a connection to the lab web browser. 

We will first define the URL from which the data can be downloaded, by concatenating the URL fo the course with the path to our dataset.

To concatenate paths, it is *recommended* to use the **R** command `file.path()`.

<small>
```{r}
url.course <- "http://jvanheld.github.io/stats_avec_RStudio_EBA/"
url.data <- file.path(url.course, "practicals", "02_peak-calling", "data")
```
</small>


## Loading a data table

**R** enables to download data directly from the Web. 

Load counts per window in chip sample.



```{r}
## Define URL of the ChIP file
chip.bedg.file <- file.path(url.data, "FNR_200bp.bedg")

## Load the file content in an R data.frame
chip.bedg <- read.table(chip.bedg.file)

## Set column names
names(chip.bedg) <- c("chrom", "start", "end","counts")
```

## Exploring a data frame: dim()

Before anything else, let us inspect the size of the data frame, in order to check that it was properly lodaded.

```{r}
dim(chip.bedg)
```


## Checking the n first rows: head()

The function `head()` displays the first rows of a table.

```{r}
head(chip.bedg, n = 5)
```

## Checking the n last rows: tail()

The function `tail()` displays the first rows of a table.

```{r}
tail(chip.bedg, n = 5)
```

## Viewing a table

The function `View()` displays the full table in a user-friendly mode.

```{r eval=FALSE}
View(chip.bedg)
```

## Selecting arbitrary rows

```{r}
chip.bedg[100:105,] 
```


## Selecting arbitrary columns

```{r}
chip.bedg[100:105, 2] 
chip.bedg[100:105, "start"] 
chip.bedg[100:105, c("start", "counts")] 
```

## Adding columns

We can add columns with the result of computations from other columns.

```{r}
chip.bedg$midpos <- (chip.bedg$start + chip.bedg$end)/2
head(chip.bedg)
```


## Plotting a density profile

We can readily print a plot with the counts per window.

```{r}
plot(chip.bedg[, c("midpos", "counts")], type="h")
```

## Plotting a density profile

Let us improve the plot

```{r}
plot(chip.bedg[, c("midpos", "counts")], type="h", 
     col="green", xlab="Genomic position (200bp windows)", 
     ylab= "Read counts",
     main="FNR ChIP-seq")
```


## Exercise: exploring the background

We already loaded the count table for the FNR ChIP counts per window. 

The background level will be estimated by loading counts per window in a genomic input sample.
These counts are available in the same directory a file named `input_200bp.bedg`

1. Load the counts per window in the input sample (genome sequencing).
2. Plot the density profile of the input
3. Compare  chip-seq and input density profiles
4. Compare counts per window between chip-seq and input

## Solution: loading the input counts per window

```{r}
## Define URL of the input file
input.bedg.file <- file.path(url.data, "input_200bp.bedg")

## Load the file content in an R data.frame
input.bedg <- read.table(input.bedg.file)

## Set column names
names(input.bedg) <- c("chrom", "start", "end","counts")

```

## Solution: plotting the input density profile

```{r}
## Compute middle positions per window
input.bedg$midpos <- (input.bedg$start + input.bedg$end)/2

plot(input.bedg[, c("midpos", "counts")], type="h", 
     col="red", xlab="Genomic position (200bp windows)", 
     ylab= "Read counts",
     main="Background (genomic input)")
```

## Solution: comparing chip-seq and background density profiles

```{r}
par(mfrow=c(2,1)) ## Draw two panels on top of each other
plot(chip.bedg[, c("midpos", "counts")], type="h", 
     col="green", xlab="Genomic position (200bp windows)", 
     ylab= "Read counts",
     main="FNR ChIP-seq")
plot(input.bedg[, c("midpos", "counts")], type="h", 
     col="red", xlab="Genomic position (200bp windows)", 
     ylab= "Read counts",
     main="Background (genomic input)")
par(mfrow=c(1,1)) ## Reset default mode

```

## Solution: comparing counts per window between chip-seq and input

```{r}
plot(input.bedg$counts, chip.bedg$counts, col="darkviolet",
     xlab="Genomic input", ylab="FNR ChIP-seq",
     main="Read counts per 200bp window")
```

## Solution: comparing counts per window between chip-seq and input

In order to better highight the dynamic range, we can use a log-based representation

```{r}
plot(input.bedg$counts, chip.bedg$counts, col="darkviolet",
     xlab="Genomic input", ylab="FNR ChIP-seq",
     main="Read counts per 200bp window",
     log="xy")
grid() ## add a grid
```

## Questions

- On the ChIP-seq versus input plot, how would you define  peaks ?
- Where would you place the limit between peaks and background fluctuations ?

## Exercises

1. Think about further drawing modes to improve your perception of the differences between signal and background.
2. We will formulate (together) a reasoning path to compute a p-value for each peak.

